### PR TITLE
Pipelines v3

### PR DESCRIPTION
**Pull request by ZachGoldberg**
# :balloon: Gruntwork Pipelines v3.0.0 Release Notes

Pipelines version 3 introduces several new features which include a number of breaking changes.  The migration guide below should take less than 5 minutes for most teams to complete, but please do make sure to follow it closely.  We're providing a checklist here to help teams ensure that steps are not skipped.

- [ ] Install and configure the Gruntwork.io GitHub App
  - [ ] (Alternative) Add `actions:read` permission to PIPELINES_READ_TOKEN
  - [ ] (Alternative) (Enterprise Only) Add PR_CREATE_TOKEN to delegated repository workflows
- [ ] Update to Terragrunt v0.67.16 or later 
- [ ] (If necessary) Add new actions to GitHub Organization allow list settings
- [ ] (Enterprise Only) Install and activate Drift Detection

# Migration Guide
<details>
<summary><h2>Install and configure the Gruntwork.io GitHub App</h2></summary>

Pipelines now uses the Gruntwork.io GitHub App for workflow permissions, which simplifies permission management. Follow the [GitHub App Installation Instructions](https://docs.gruntwork.io/foundations/github-app/setup) to install the Gruntwork.io GitHub App and add your `infrastructure-live-root` repository to your account.

Using the Gruntwork.io GitHub App is recommended and provides full feature support for pipelines workflows. Alternatively, if you do not wish to install the Gruntwork.io GitHub App you can modify your existing workflows to continue using GitHub secrets, although some features will not be available. Read more about App only features [here](https://docs.gruntwork.io/foundations/github-app/availability#app-only-features).
<details>
<summary><h3>Alternative steps</h3></summary>

1. Add `actions: read` permissions

Customers explicitly list permissions that Gruntwork Pipelines workflows has by default in the pipelines workflow files in `.github/workflows` inside their infrastructure repositories.  Pipelines v3 now requires `actions: read` permissions in order to introspect its own runs and provide more helpful logging capabilities.

Customers should make the one-line change to add the `actions: read` permission in the following files <b>in every repository that uses Gruntwork pipelines</b> (including root, access control and delegated repos).  Note, most repositories will have only 1 or two of these workflow files, in which case update what is present and don't worry about the others.

* `.github/workflows/pipelines.yml`
* `.github/workflows/pipelines-drift-detection.yml`
* `.github/workflows/pipelines-root.yml`
* `.github/workflows/pipelines-unlock.yml`

### Old Permissions
```yml
permissions:
  id-token: write
  contents: write
  pull-requests: write
```
### New Permissions
```yml
permissions:
  id-token: write
  contents: write
  pull-requests: write
  actions: read
```

2. (Enterprise only) Add PR_CREATE_TOKEN to delegated repository workflows.

Delegated repositories that are not using the Gruntwork.io GitHub App will need to add  `PR_CREATE_TOKEN: ${{ github.token }}` to the secrets in their workflow files:

### Old Secrets
```yml
secrets:
  PIPELINES_READ_TOKEN: ${{ secrets.PIPELINES_READ_TOKEN }}
```
### New Secrets
```yml
secrets:
  PIPELINES_READ_TOKEN: ${{ secrets.PIPELINES_READ_TOKEN }}
  PR_CREATE_TOKEN: ${{ github.token }}
```

</details>

</details>
<details>
<summary><h2>Update to Terragrunt v0.67.16</h2></summary>

The latest pipelines works best with the latest version of Terragrunt.  Recent versions of Terragrunt have much improved logging, performance and correctness improvements in `run-all` scenarios with pipelines.
</details>
<details>
<summary><h2>Allowlist Actions</h2></summary>
<i>This is only for customers who only allow GitHub actions to run if they are on an <a href="https://docs.github.com/en/organizations/managing-organization-settings/disabling-or-limiting-github-actions-for-your-organization#allowing-select-actions-and-reusable-workflows-to-run">explicit allowlist</a></i>

<h3>New actions to add</h3>

* `gruntwork-io/pipelines-credentials`
* `gruntwork-io/pipelines-actions/.github/actions/pipelines-drift-detection-consolidate-jobs`
* `gruntwork-io/pipelines-actions/.github/actions/pipelines-drift-detection-determine-units`
* `gruntwork-io/pipelines-actions/.github/actions/pipelines-drift-detection-determine-drift`
* `gruntwork-io/pipelines-actions/.github/actions/pipelines-new-pr-action`
* `gruntwork-io/pipelines-actions/.github/actions/pipelines-get-job-logs-url`
</details>
<details>
<summary><h2>(Enterprise Only) Install and activate Drift Detection</h2></summary>
Pipelines Drift Detection can be installed in your repositories by adding a new workflow file:

1. Create a new file at `.github/workflows/pipelines-drift-detection.yml`
2. Add the following content to the file

```yml
name: Pipelines Drift Detection
run-name: "[GWP]: Pipelines Drift Detection"
on:
  # Uncomment to enable scheduled Drift Detection
  # schedule:
  #  - cron: '15 12 * * 1'
  workflow_dispatch:
    inputs:
      path:
        description: (Optional) Path to filter units e.g. "./management/*"
        type: string
      branch-name:
        description: (Optional) branch name to open Drift Detection PRs with
        default: drift-detection
        type: string
permissions:
  id-token: write

jobs:
  GruntworkPipelines:
    uses: gruntwork-io/pipelines-workflows/.github/workflows/pipelines-drift-detection.yml@v3
    with:
      path: ${{ inputs.path }}
      branch-name: ${{ inputs.branch-name }}
```

</details>

# New Features
- [Pipelines as a GitHub App](https://docs.gruntwork.io/foundations/github-app/)
- [Pipelines Drift Detection](https://docs.gruntwork.io/pipelines/maintain/drift-detection) (Enterprise Only)
- Improved Account Factory Bootstrap Customization Hooks
- Pipelines Log Link Improvements

### CODE REVIEW

File: .github/workflows/pipelines-drift-detection.yml [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+jobs:
+  determine_units:
+    name: Detect Infrastructure Drift
+    runs-on: ${{ fromJSON(inputs.runner) }}
+    outputs:
+      units: ${{ steps.determine-units.outputs.units }}
+    steps:
+      - name: Fetch Gruntwork Read Token
+        id: pipelines-gruntwork-read-token
+        uses: gruntwork-io/pipelines-credentials@main
```
Comment (yhakbar) [inline .github/workflows/pipelines-drift-detection.yml:50]:
```suggestion
        uses: gruntwork-io/pipelines-credentials@v1
```

File: .github/workflows/pipelines-drift-detection.yml [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        with:
+          PIPELINES_TOKEN_PATH: "pipelines-read/gruntwork-io"
+          FALLBACK_TOKEN: ${{ secrets.PIPELINES_READ_TOKEN }}
+          api_base_url: ${{ inputs.api_base_url }}
+
+
+      - name: Fetch Org Read Token
+        id: pipelines-customer-org-read-token
+        uses: gruntwork-io/pipelines-credentials@main
```
Comment (yhakbar) [inline .github/workflows/pipelines-drift-detection.yml:59]:
```suggestion
        uses: gruntwork-io/pipelines-credentials@v1
```

File: .github/workflows/pipelines-drift-detection.yml [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+      fail-fast: false
+      matrix:
+        unit: ${{ fromJSON(needs.determine_units.outputs.units) }}
+    env:
+      JOB_NAME: Detect Drift in ${{ matrix.unit.path }}
+    name: Detect Drift in ${{ matrix.unit.path }}
+    steps:
+      - name: Fetch Gruntwork Read Token
+        id: pipelines-gruntwork-read-token
+        uses: gruntwork-io/pipelines-credentials@main
```
Comment (yhakbar) [inline .github/workflows/pipelines-drift-detection.yml:97]:
```suggestion
        uses: gruntwork-io/pipelines-credentials@v1
```

File: .github/workflows/pipelines-drift-detection.yml [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        with:
+          PIPELINES_TOKEN_PATH: "pipelines-read/gruntwork-io"
+          FALLBACK_TOKEN: ${{ secrets.PIPELINES_READ_TOKEN }}
+          api_base_url: ${{ inputs.api_base_url }}
+
+      - name: Fetch Org Read Token
+        id: pipelines-customer-org-read-token
+        uses: gruntwork-io/pipelines-credentials@main
```
Comment (yhakbar) [inline .github/workflows/pipelines-drift-detection.yml:105]:
```suggestion
        uses: gruntwork-io/pipelines-credentials@v1
```

File: .github/workflows/pipelines-drift-detection.yml [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+          JOB_NAME: ${{ env.JOB_NAME }}
+          STEP_NAME: "Execute Drift Detection"
+
+  consolidate_jobs:
+    runs-on: ${{ fromJSON(inputs.runner) }}
+    needs: pipelines_drift_detection
+    steps:
+      - name: Fetch Gruntwork Read Token
+        id: pipelines-gruntwork-read-token
+        uses: gruntwork-io/pipelines-credentials@main
```
Comment (yhakbar) [inline .github/workflows/pipelines-drift-detection.yml:143]:
```suggestion
        uses: gruntwork-io/pipelines-credentials@v1
```

File: .github/workflows/pipelines-drift-detection.yml [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+        with:
+          PIPELINES_TOKEN_PATH: "pipelines-read/gruntwork-io"
+          FALLBACK_TOKEN: ${{ secrets.PIPELINES_READ_TOKEN }}
+          api_base_url: ${{ inputs.api_base_url }}
+
+      - name: Fetch Org Read Token
+        id: pipelines-customer-org-read-token
+        uses: gruntwork-io/pipelines-credentials@main
```
Comment (yhakbar) [inline .github/workflows/pipelines-drift-detection.yml:151]:
```suggestion
        uses: gruntwork-io/pipelines-credentials@v1
```

File: .github/workflows/pipelines-drift-detection.yml [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+        with:
+          PIPELINES_TOKEN_PATH: pipelines-read/${{ github.repository_owner }}
+          FALLBACK_TOKEN: ${{ secrets.PIPELINES_READ_TOKEN }}
+          api_base_url: ${{ inputs.api_base_url }}
+
+      - name: Fetch Create PR Token
+        id: pipelines-propose-infra-change-token
+        uses: gruntwork-io/pipelines-credentials@main
```
Comment (yhakbar) [inline .github/workflows/pipelines-drift-detection.yml:159]:
```suggestion
        uses: gruntwork-io/pipelines-credentials@v1
```

File: .github/workflows/pipelines-root.yml [hunk 8]
```diff
@@ -49,14 +53,46 @@ jobs:
     name: Detect Infrastructure Changes
     runs-on: ${{ fromJSON(inputs.runner) }}
     steps:
+      - name: Fetch Gruntwork Read Token
+        id: pipelines-gruntwork-read-token
+        uses: gruntwork-io/pipelines-credentials@main
```
Comment (yhakbar) [inline .github/workflows/pipelines-root.yml:58]:
```suggestion
        uses: gruntwork-io/pipelines-credentials@v1
```

File: .github/workflows/pipelines-root.yml [hunk 9]
```diff
@@ -49,14 +53,46 @@ jobs:
...
+        with:
+          PIPELINES_TOKEN_PATH: "pipelines-read/gruntwork-io"
+          FALLBACK_TOKEN: ${{ secrets.PIPELINES_READ_TOKEN }}
+          api_base_url: ${{ inputs.api_base_url }}
+
+      - name: Fetch Org Read Token
+        id: pipelines-customer-org-read-token
+        uses: gruntwork-io/pipelines-credentials@main
```
Comment (yhakbar) [inline .github/workflows/pipelines-root.yml:66]:
```suggestion
        uses: gruntwork-io/pipelines-credentials@v1
```

File: .github/workflows/pipelines-root.yml [hunk 10]
```diff
@@ -49,14 +53,46 @@ jobs:
...
+        with:
+          PIPELINES_TOKEN_PATH: pipelines-read/${{ github.repository_owner }}
+          FALLBACK_TOKEN: ${{ secrets.PIPELINES_READ_TOKEN }}
+          api_base_url: ${{ inputs.api_base_url }}
+
+      - name: Fetch Infra Root Write Token
+        id: pipelines-infra-root-write-token
+        uses: gruntwork-io/pipelines-credentials@main
```
Comment (yhakbar) [inline .github/workflows/pipelines-root.yml:74]:
```suggestion
        uses: gruntwork-io/pipelines-credentials@v1
```

File: .github/workflows/pipelines-root.yml [hunk 11]
```diff
@@ -49,14 +53,46 @@ jobs:
...
+        with:
+          PIPELINES_TOKEN_PATH: infra-root-write/${{ github.repository_owner }}
+          FALLBACK_TOKEN: ${{ secrets.INFRA_ROOT_WRITE_TOKEN }}
+          api_base_url: ${{ inputs.api_base_url }}
+
+      - name: Fetch Org Repo Admin Token
+        id: pipelines-org-repo-admin-token
+        uses: gruntwork-io/pipelines-credentials@main
```
Comment (yhakbar) [inline .github/workflows/pipelines-root.yml:82]:
```suggestion
        uses: gruntwork-io/pipelines-credentials@v1
```

File: .github/workflows/pipelines-root.yml [hunk 12]
```diff
@@ -107,25 +151,51 @@ jobs:
       matrix:
         jobs: ${{ fromJson(needs.pipelines_orchestrate.outputs.pipelines_jobs) }}
     steps:
+      - name: Fetch Gruntwork Read Token
+        id: pipelines-gruntwork-read-token
+        uses: gruntwork-io/pipelines-credentials@main
```
Comment (yhakbar) [inline .github/workflows/pipelines-root.yml:156]:
```suggestion
        uses: gruntwork-io/pipelines-credentials@v1
```

File: .github/workflows/pipelines-root.yml [hunk 13]
```diff
@@ -107,25 +151,51 @@ jobs:
...
+        with:
+          PIPELINES_TOKEN_PATH: "pipelines-read/gruntwork-io"
+          FALLBACK_TOKEN: ${{ secrets.PIPELINES_READ_TOKEN }}
+          api_base_url: ${{ inputs.api_base_url }}
+
+      - name: Fetch Org Read Token
+        id: pipelines-customer-org-read-token
+        uses: gruntwork-io/pipelines-credentials@main
```
Comment (yhakbar) [inline .github/workflows/pipelines-root.yml:164]:
```suggestion
        uses: gruntwork-io/pipelines-credentials@v1
```

File: .github/workflows/pipelines-root.yml [hunk 14]
```diff
@@ -107,25 +151,51 @@ jobs:
...
+        with:
+          PIPELINES_TOKEN_PATH: pipelines-read/${{ github.repository_owner }}
+          FALLBACK_TOKEN: ${{ secrets.PIPELINES_READ_TOKEN }}
+          api_base_url: ${{ inputs.api_base_url }}
+
+      - name: Fetch Infra Root Write Token
+        id: pipelines-infra-root-write-token
+        uses: gruntwork-io/pipelines-credentials@main
```
Comment (yhakbar) [inline .github/workflows/pipelines-root.yml:172]:
```suggestion
        uses: gruntwork-io/pipelines-credentials@v1
```

File: .github/workflows/pipelines-root.yml [hunk 15]
```diff
@@ -238,32 +342,59 @@ jobs:
       matrix:
         jobs: ${{ fromJson(needs.pipelines_orchestrate.outputs.pipelines_jobs)[0].NewAccounts }}
     steps:
+      - name: Fetch Gruntwork Read Token
+        id: pipelines-gruntwork-read-token
+        uses: gruntwork-io/pipelines-credentials@main
```
Comment (yhakbar) [inline .github/workflows/pipelines-root.yml:347]:
```suggestion
        uses: gruntwork-io/pipelines-credentials@v1
```

File: .github/workflows/pipelines-root.yml [hunk 16]
```diff
@@ -238,32 +342,59 @@ jobs:
...
+        with:
+          PIPELINES_TOKEN_PATH: "pipelines-read/gruntwork-io"
+          FALLBACK_TOKEN: ${{ secrets.PIPELINES_READ_TOKEN }}
+          api_base_url: ${{ inputs.api_base_url }}
+
+      - name: Fetch Org Read Token
+        id: pipelines-customer-org-read-token
+        uses: gruntwork-io/pipelines-credentials@main
```
Comment (yhakbar) [inline .github/workflows/pipelines-root.yml:355]:
```suggestion
        uses: gruntwork-io/pipelines-credentials@v1
```

File: .github/workflows/pipelines-drift-detection.yml [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+        uses: ./pipelines-actions/.github/actions/pipelines-drift-detection-determine-drift
+        with:
+          PIPELINES_GRUNTWORK_READ_TOKEN: ${{ steps.pipelines-gruntwork-read-token.outputs.PIPELINES_TOKEN }}
+          PIPELINES_CUSTOMER_ORG_READ_TOKEN: ${{ steps.pipelines-customer-org-read-token.outputs.PIPELINES_TOKEN }}
+          UNIT_ID: ${{ matrix.unit.id }}
+          UNIT_PATH: ${{ matrix.unit.path }}
+          JOB_NAME: ${{ env.JOB_NAME }}
+          STEP_NAME: "Execute Drift Detection"
+
+  consolidate_jobs:
```
Comment (yhakbar) [inline .github/workflows/pipelines-drift-detection.yml:137]:
```suggestion
  consolidate_jobs:
  name: Consolidate Jobs
```

File: .github/workflows/pipelines-unlock.yml [hunk 18]
```diff
@@ -223,25 +262,43 @@ jobs:
...
+        id: pipelines-gruntwork-read-token
+        uses: gruntwork-io/pipelines-credentials@main
+        with:
+          PIPELINES_TOKEN_PATH: "pipelines-read/gruntwork-io"
+          FALLBACK_TOKEN: ${{ secrets.PIPELINES_READ_TOKEN }}
+          api_base_url: ${{ inputs.api_base_url }}
+
+      - name: Fetch Org Read Token
+        id: pipelines-customer-org-read-token
+        uses: gruntwork-io/pipelines-credentials@main
```
Comment (yhakbar) [inline .github/workflows/pipelines-unlock.yml:275]:
```suggestion
        uses: gruntwork-io/pipelines-credentials@v1
```

**Review by yhakbar — CHANGES_REQUESTED** [review]

File: .github/workflows/pipelines-drift-detection.yml [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+      api_base_url:
+        type: string
+        default: "https://api.prod.app.gruntwork.io/api/v1"
+    secrets:
+        PIPELINES_READ_TOKEN:
+          required: false
+        PR_CREATE_TOKEN:
+          required: false
+env:
+  PIPELINES_CLI_VERSION: v0.29.0-rc17
```
Comment (Resonance1584) [inline .github/workflows/pipelines-drift-detection.yml:39]:
```suggestion
  PIPELINES_CLI_VERSION: v0.29.0
```

File: .github/workflows/pipelines-drift-detection.yml [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+  PIPELINES_ACTIONS_VERSION: v3-rc1
```
Comment (Resonance1584) [inline .github/workflows/pipelines-drift-detection.yml:40]:
```suggestion
  PIPELINES_ACTIONS_VERSION: v3
```

File: .github/workflows/pipelines-root.yml [hunk 21]
```diff
@@ -30,8 +34,8 @@ on:
         required: false
 
 env:
-  PIPELINES_CLI_VERSION: v0.28.0
-  PIPELINES_ACTIONS_VERSION: main
+  PIPELINES_CLI_VERSION: v0.29.0-rc17
```
Comment (Resonance1584) [inline .github/workflows/pipelines-root.yml:37]:
```suggestion
  PIPELINES_CLI_VERSION: v0.29.0
```

File: .github/workflows/pipelines-root.yml [hunk 22]
```diff
@@ -30,8 +34,8 @@ on:
...
+  PIPELINES_ACTIONS_VERSION: v3-rc1
```
Comment (Resonance1584) [inline .github/workflows/pipelines-root.yml:38]:
```suggestion
  PIPELINES_ACTIONS_VERSION: v3
```

File: .github/workflows/pipelines-unlock.yml [hunk 23]
```diff
@@ -38,34 +41,52 @@ on:
         PIPELINES_READ_TOKEN:
           required: true
 env:
-  PIPELINES_CLI_VERSION: v0.28.0
-  PIPELINES_ACTIONS_VERSION: main
+  PIPELINES_CLI_VERSION: v0.29.0-rc17
```
Comment (Resonance1584) [inline .github/workflows/pipelines-unlock.yml:44]:
```suggestion
  PIPELINES_CLI_VERSION: v0.29.0
```

File: .github/workflows/pipelines-unlock.yml [hunk 24]
```diff
@@ -38,34 +41,52 @@ on:
...
+  PIPELINES_ACTIONS_VERSION: v3-rc1
```
Comment (Resonance1584) [inline .github/workflows/pipelines-unlock.yml:45]:
```suggestion
  PIPELINES_ACTIONS_VERSION: v3
```

File: .github/workflows/pipelines.yml [hunk 25]
```diff
@@ -21,13 +21,19 @@ on:
...
     secrets:
       PIPELINES_READ_TOKEN:
         required: true
+      PR_CREATE_TOKEN:
+          required: false
 
 env:
-  PIPELINES_CLI_VERSION: v0.28.0
-  PIPELINES_ACTIONS_VERSION: main
+  PIPELINES_CLI_VERSION: v0.29.0-rc17
```
Comment (Resonance1584) [inline .github/workflows/pipelines.yml:35]:
```suggestion
  PIPELINES_CLI_VERSION: v0.29.0
```

File: .github/workflows/pipelines.yml [hunk 26]
```diff
@@ -21,13 +21,19 @@ on:
...
+  PIPELINES_ACTIONS_VERSION: v3-rc1
```
Comment (Resonance1584) [inline .github/workflows/pipelines.yml:36]:
```suggestion
  PIPELINES_ACTIONS_VERSION: v3
```

File: .github/workflows/pipelines-drift-detection.yml [hunk 27]
```diff
@@ -0,0 +1,195 @@
+name: Pipelines
+run-name: Drift Detection
+on:
+  workflow_call:
+    inputs:
+      # This field can be overriden to customize the runner used for pipelines
+      # workflows.
+      #
```
Comment (oredavids) [inline .github/workflows/pipelines-drift-detection.yml:8]: This documentation should be on the `runner` input. i.e move the `path` input

**Review by oredavids — DISMISSED** [review]

File: .github/workflows/pipelines-root.yml [hunk 28]
```diff
@@ -205,6 +296,15 @@ jobs:
...
           deploy_branch_name: ${{ steps.gruntwork_context.outputs.deploy_branch_name }}
 
+      - name: Get Logs URL
+        id: get_logs_url
+        uses: ./pipelines-actions/.github/actions/pipelines-get-job-logs-url
+        if: always()
+        with:
+          PIPELINES_CUSTOMER_ORG_READ_TOKEN: ${{ steps.pipelines-customer-org-read-token.outputs.PIPELINES_TOKEN }}
+          job_name: ${{ env.JOB_NAME }}
+          step_name_prefix: "${{ steps.gruntwork_context.outputs.action == 'TERRAGRUNT_EXECUTE' && '[TerragruntExecute]:\ Authenticate with AWS and then Invoke Terragrunt' || (steps.gruntwork_context.outputs.action == 'BASELINE_ACCOUNT' && 'Run core accounts baselines' || '[ProvisionAccount]:\ Provision New Account') }}"
```
Comment (yhakbar) [inline .github/workflows/pipelines-root.yml:306]: NIT: This should be refactored out at some point. It's hard how this resolves on first inspection.

**Review by yhakbar — APPROVED** [review]

**Pull request merged**